src/System.Windows.Forms/src/System/Windows/Forms/NativeWindow.cs

```
@@ -500,8 +500,6 @@ public void DefWndProc(ref Message m)
             {
                 if (_priorWindowProcHandle == IntPtr.Zero)
                 {
-                    Debug.Fail($"Can't find a default window procedure for message {m} on class {GetType().Name}");
```

Why are we removing this?

---

Because it can be hit by a unit test. The `Debug.Fail` should be removed rather than ignored because there is code below that should be tested and can be reached

---

Like said elsewhere I think you have to carefully look at how you are reaching the code. It is very easy to reach basically everything via unit test, but that doesn't mean it can come up in a valid scenario.

This Debug Assert is there to make sure the normal code paths are coded properly. The logic afterwards is a "best effort panic cleanup" and not something you would want to become a "normal code path". This codepath should never execute under normal operations, if you want to test abnormal operations the test can easily use a NoAssertContext.

I'm concerned that the removal of asserts without looking into their meaning reduces future code quality.

---

I believe you can repro this Assert failure in the following:
```cs
var window = new NativeWindow();
var m = new Message();
window.DefWndProc(ref m);
```

---

Yes and this is a perfect example of wrong usage of an internal API, anyone doing this call should fail the assert and be fixed immediately. APIs have semantics which are supposed to be honored by callers, testing random calls with random nonsense arguments is not what unit tests are about.

NativeWindow is a very fragile class, getting something wrong here risks the stability of the whole process. See #2380 for a case where things go really bad, so its not just theoretical. If WinForms starts to run through WndProcs with wrong callbacks you are at risk of introducing really bad bugs.

You want to catch anyone accidently going through this code path in testing. You can't get full test coverage of not-yet-written classes so having asserts in here is the best way of avoiding that particular code path.

If you had the option you would definitely throw exceptions, but you do not have this option when things go bad because the caller of WndProc (usually) won't be in a position to catch exceptions. So your only choice is to do a "best effort" and clean up as much as you can (or just crash the process with a FailFast). That doesn't mean you don't want to detect those bugs.